File: src/main/java/github/saukiya/sxattribute/data/attribute/sub/other/JSAttribute.java
package github.saukiya.sxattribute.data.attribute.sub.other;

import github.saukiya.sxattribute.SXAttribute;
import github.saukiya.sxattribute.data.attribute.AttributeType;
import github.saukiya.sxattribute.data.attribute.SubAttribute;
import github.saukiya.sxattribute.data.eventdata.EventData;
import lombok.Getter;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;

import javax.script.*;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Saukiya
 */
@Getter
public class JSAttribute extends SubAttribute {
    private static final Object INVOCATION_FAILED = new Object();
    private final ScriptEngine script;
    private final Invocable invocable;
    /** 脚本熔断状态；失败后所有属性回调退化为无效果，避免持续影响事件链。 */
    private volatile boolean active = true;
    private volatile boolean failureLogged;

    @SuppressWarnings("unchecked")
    public JSAttribute(String name, ScriptEngine script) {
        super(name, SXAttribute.getInst(), script.get("valuesLength") != null ? (int) script.get("valuesLength") : 0,
                ((List<AttributeType>) script.get("types")).toArray(new AttributeType[0]));
        this.script = script;
        invocable = (Invocable) script;
        script.put(name, this);
    }

    @Override
    public void setLength(int length) {
        super.setLength(length);
        script.put("valuesLength", length);
    }

    @Override
    public File getConfigFile() {
        return new File(SXAttribute.getInst().getDataFolder(), "Attribute" + File.separator
                + getPlugin().getName() + File.separator + getName() + "_JS.yml");
    }

    /** 加载脚本属性；先迁移已知坏配置，单个脚本失败时只禁用该属性而不阻断插件启动。 */
    @Override
    public void onEnable() {
        script.put("config", config());
        migrateBrokenTickScript();
        active = true;
        failureLogged = false;
        invoke("onEnable", null);
    }

    /**
     * 修复 3.9.x/4.0.0-beta.1 已生成的 Tick1 默认脚本。
     * <p>
     * 旧脚本在三个 JavaScript 语句之间漏写分号，Nashorn 会把 {@code )player} 解析为非法语法并中止
     * 整个插件启动。这里只匹配内置 Tick1 的特征文本，绝不改写服主自行编写的其它脚本。
     */
    private void migrateBrokenTickScript() {
        String path = "List.Tick1.Continued";
        List<String> scripts = new ArrayList<>(config().getStringList(path));
        if (repairLegacyTickScript(scripts)) {
            config().set(path, scripts);
            SXAttribute.getInst().getLogger().info("Migrated invalid built-in JSAttribute Tick1 script.");
        }
    }

    static boolean repairLegacyTickScript(List<String> scripts) {
        boolean changed = false;
        for (int index = 0; index < scripts.size(); index++) {
            String value = scripts.get(index);
            if (!value.contains("js.Tick1Boo") || !value.contains("Particle.SPELL_WITCH")
                    || !value.contains("Particle.SPELL_MOB")) {
                continue;
            }
            String repaired = value
                    .replace("js.drawRound(js.Tick1Angle, 2)player", "js.drawRound(js.Tick1Angle, 2);player")
                    .replace(", 0)player.getWorld().spawnParticle(Particle.SPELL_MOB",
                            ", 0);player.getWorld().spawnParticle(Particle.SPELL_MOB");
            if (!repaired.trim().endsWith(";")) repaired += ";";
            if (!repaired.equals(value)) {
                scripts.set(index, repaired);
                changed = true;
            }
        }
        return changed;
    }

    @Override
    public void onReLoad() {
        script.put("config", config());
        active = true;
        failureLogged = false;
        migrateBrokenTickScript();
        invoke("onReLoad", null);
    }

    @Override
    public void onDisable() {
        invoke("onDisable", null);
        active = false;
    }

    @Override
    protected YamlConfiguration defaultConfig(YamlConfiguration config) {
        Object result = invoke("defaultConfig", config, config);
        return result instanceof YamlConfiguration ? (YamlConfiguration) result : config;
    }

    @Override
    public void eventMethod(double[] values, EventData eventData) {
        invoke("eventMethod", null, values, eventData);
    }

    @Override
    public Object getPlaceholder(double[] values, Player player, String string) {
        return invoke("getPlaceholder", null, values, player, string);
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<String> getPlaceholders() {
        try {
            Object object = invoke("getPlaceholders", null);
            if (object instanceof List) {
                return (List<String>) object;
            } else if (object instanceof Bindings) {
                Bindings som = (Bindings) object;
                List<String> list = new ArrayList<>();
                int i = 0;
                while (som.get(String.valueOf(i)) != null) {
                    list.add(som.get(String.valueOf(i++)).toString());
                }
                return list;
            }
            return null;
        } catch (RuntimeException exception) {
            disable("getPlaceholders conversion", exception);
            return null;
        }
    }

    @Override
    public void loadAttribute(double[] values, String lore) {
        invoke("loadAttribute", null, values, lore);
    }

    @Override
    public void correct(double[] values) {
        if (invoke("correct", INVOCATION_FAILED, values) == INVOCATION_FAILED) super.correct(values);
    }

    @Override
    public double calculationCombatPower(double[] values) {
        try {
            Object result = invoke("calculationCombatPower", null, values);
            return result == null ? 0D : Double.parseDouble(result.toString());
        } catch (RuntimeException exception) {
            disable("calculationCombatPower conversion", exception);
            return 0;
        }
    }

    /**
     * 非sx调用的方法
     * js属性需要调用此方法则为 属性名.method("参数");
     *
     * @param type
     * @return
     */
    public Object method(String type) {
        return invoke("method", null, type);
    }

    /**
     * 创建一个Runnable 并塞进一个脚本
     * 用处:
     *
     * @param cs       CompiledScript
     * @param bindings Bindings
     * @return
     */
    public Runnable createRunnable(CompiledScript cs, Bindings bindings) {
        return () -> {
            if (!active) return;
            try {
                if (bindings != null) {
                    cs.eval(bindings);
                } else {
                    cs.eval();
                }
            } catch (Throwable exception) {
                rethrowFatal(exception);
                disable("scheduled script", exception);
            }
        };
    }

    /**
     * 统一执行脚本函数。缺少可选函数时返回降级值；脚本自身异常会触发熔断，但不向 Bukkit 事件链抛出。
     */
    private Object invoke(String function, Object fallback, Object... arguments) {
        if (!active) return fallback;
        try {
            return invocable.invokeFunction(function, arguments);
        } catch (NoSuchMethodException exception) {
            if (Boolean.TRUE.equals(script.get(function + "_debug"))) {
                SXAttribute.getInst().getLogger().warning("JS attribute " + getName() + " has no function " + function);
            }
            return fallback;
        } catch (Throwable exception) {
            rethrowFatal(exception);
            disable(function, exception);
            return fallback;
        }
    }

    /** 首次异常后熔断当前属性并只记录一次摘要，防止周期脚本持续刷屏。 */
    private void disable(String phase, Throwable exception) {
        active = false;
        if (failureLogged) return;
        failureLogged = true;
        SXAttribute.getInst().getLogger().severe("JS attribute " + getName() + " failed in " + phase
                + " and was disabled without affecting SX-Attribute: " + exception.getClass().getSimpleName()
                + ": " + exception.getMessage());
    }

    static boolean isFatalFailure(Throwable exception) {
        return exception instanceof VirtualMachineError;
    }

    private static void rethrowFatal(Throwable exception) {
        if (exception instanceof VirtualMachineError) throw (VirtualMachineError) exception;
    }

    /**
     * 向bindings添加数据
     * js脚本中无法对Bindings进行操作，因为本身为 脚本对象镜像 ScriptObjectMirror
     *
     * @param bindings create -> Engine.createBindings()
     * @param key
     * @param value
     */
    public Bindings put(Bindings bindings, String key, Object value) {
        bindings.put(key, value);
        return bindings;
    }
}

File: src/main/resources/Attribute/JavaScript/JSAttribute.js
// var jsManager = Java.type("javax.script.ScriptEngineManager").class.newInstance();
var jsManager = new Packages.javax.script.ScriptEngineManager();
// var Player = Java.type("org.bukkit.entity.Player").class;
var Player = Packages.org.bukkit.entity.Player.class;
var Particle = Packages.org.bukkit.Particle;

var valuesLength = 0;
// 申明个鬼 直接types
var types = Arrays.asList(SXAttributeType.UPDATE);

var JSRunnableList = [];

var js = this;

// 已含有变量:
// -对象:
// 文件名 - JSAttribute java
//
// -静态类:
// SXAttributeType
// SXAttribute
// Bukkit
// Arrays
// API
//

/**
 * 添加String类方法 startsWith
 */
if (typeof String.prototype.startsWith != 'function') {
    String.prototype.startsWith = function (prefix) {
        return this.slice(0, prefix.length) === prefix;
    };
}

/**
 * 向JSRunnableList添加方法 getRunnable
 */
if (typeof JSRunnableList.getRunnable != 'function') {
    JSRunnableList.getRunnable = function (name) {
        for (var i = 0; i < this.length; i++) {
            if (this[i].name === name) {
                return this[i];
            }
        }
        return null;
    }
}

/**
 * 创建一个与玩家关联的 binding
 */
function getBinding(player) {
    var bindings = engine.createBindings();
    JSAttribute.put(bindings, "player", player);
    JSAttribute.put(bindings, "js", js);
    JSAttribute.put(bindings, "Particle", Particle);
    return bindings;
}

/**
 * JS工厂
 *
 * 创建jsRunnable -> JSRunnableManager.create(name);
 *
 */
var JSRunnableManager = {
    engine: jsManager.getEngineByName("JavaScript"),

    create: function (name) {
        var JSRunnable = {
            cycleRun: engine.compile("JSRunnable.run();"),
            name: name,
            data: {},
            discernName: undefined,
            enabledScript: [],
            disableScript: [],
            continuedScript: [],
            combatPower: undefined,
            players: [],

            load: function () {
                var JSRunnable = this;
                // this.discernName = JSAttribute.getString("List." + name + ".DiscernName");
                this.discernName = config.getString("List." + name + ".DiscernName");
                this.data = {};
                this.enabledScript = [];
                this.disableScript = [];
                this.continuedScript = [];
                this.combatPower = config.getInt("List." + name + ".CombatPower");

                config.getStringList("List." + name + ".Enabled").forEach(function (value) {
                    JSRunnable.enabledScript.push(value.startsWith("delay") ? value : engine.compile(value));
                });
                config.getStringList("List." + name + ".Disable").forEach(function (value) {
                    JSRunnable.disableScript.push(value.startsWith("delay") ? value : engine.compile(value));
                });
                config.getStringList("List." + name + ".Continued").forEach(function (value) {
                    JSRunnable.continuedScript.push(value.startsWith("delay") ? value : engine.compile(value));
                });
                var bindings = engine.createBindings();
                JSAttribute.put(bindings, "js", js);
                JSAttribute.put(bindings, "data", this.data);
                config.getString("List." + name + ".Initialization")
                config.getStringList("List." + name + ".Initialization").forEach(function (value) {
                    engine.compile(value).eval(bindings);
                })
            },


            enabled: function (player) {
                this.players.push(player.getName());
                this.runScript(player, this.enabledScript);
            },

            disable: function (player) {
                var playerName = player.getName();
                for (var i = 0; i < this.players.length; i++) {
                    if (this.players[i] === playerName) {
                        this.players.splice(i, 1);
                    }
                }
                this.runScript(player, this.disableScript);
            },

            runScript: function (player, script) {
                var delay = 0;
                for (var scriptKey in script) {
                    var value = script[scriptKey];
                    if (typeof value == "string") {
                        delay += parseInt(value.slice(6, value.length));
                    } else {
                        Bukkit.getScheduler().runTaskLater(JSAttribute.getPlugin(), JSAttribute.createRunnable(value, JSAttribute.put(getBinding(player), "data", this.data)), delay);
                    }
                }
                return delay;
            },

            run: function () {
                var delay = 0;
                for (var scriptKey in this.continuedScript) {
                    var value = this.continuedScript[scriptKey];
                    if (typeof value == "string") {
                        delay += parseInt(value.slice(6, value.length));
                    } else {
                        for (var playerKey in this.players) {
                            var playerName = this.players[playerKey];
                            var player = Bukkit.getPlayerExact(playerName);
                            if (player != null) {
                                Bukkit.getScheduler().runTaskLater(JSAttribute.getPlugin(), JSAttribute.createRunnable(value, JSAttribute.put(getBinding(player), "data", this.data)), delay);
                            } else {
                                for (var i = 0; i < this.players.length; i++) {
                                    if (this.players[i] === playerName) {
                                        this.players.splice(i, 1);
                                    }
                                }
                            }

                        }
                    }
                }
                Bukkit.getScheduler().runTaskLater(JSAttribute.getPlugin(), JSAttribute.createRunnable(this.cycleRun, this.cycleBinding), delay === 0 ? 20 : delay);
            }
        };
        JSRunnable.cycleBinding = this.engine.createBindings();
        JSAttribute.put(JSRunnable.cycleBinding, "JSRunnable", JSRunnable);

        /**
         * 添加 JSRunnable.players 方法 contains
         */
        if (typeof JSRunnable.players.contains != 'function') {
            JSRunnable.players.contains = function (object) {
                for (var i = 0; i < this.length; i++) {
                    if (this[i] === object) {
                        return true;
                    }
                }
                return false;
            }
        }
        JSRunnable.load();
        JSRunnable.run();

        return JSRunnable;
    }

}

/**
 * 以下为sx调用方法
 *
 * 注意: debug开启方法为方法名+debug（例: eventMethod_debug = true)
 */

/**
 * 创建默认属性设置
 *
 * @param config YamlConfiguration
 * @returns YamlConfiguration
 */
// defaultConfig_debug = true
function defaultConfig(config) {
    config.set("List.Tick1.DiscernName", "耀眼光芒");
    config.set("List.Tick1.Enabled", Arrays.asList("js.Tick1 = js.Tick1 != null ? js.Tick1 : 1;js.Tick1Angle = js.Tick1Angle != null ? js.Tick1Angle : 0;js.Tick1Boo = js.Tick1Boo != null ? js.Tick1Boo : true;if (typeof js.drawRound != \"function\") {js.drawRound = function(angle, radians) {var radiansCircle = angle / 180.0 * Math.PI;return [Math.cos(radiansCircle) * radians, Math.sin(radiansCircle) * radians];}}player.getWorld().spawnParticle(Particle.REDSTONE, player.getLocation().clone().add(0, 1, 0), 30, 0.5, 0.5, 0.5, 1);"));
    config.set("List.Tick1.Disable", Arrays.asList("player.getWorld().spawnParticle(Particle.EXPLOSION_LARGE, player.getLocation().clone().add(0, 1, 0), 3, 0.5, 0.5, 0.5, 0);"));
    config.set("List.Tick1.Continued", Arrays.asList("delay 1", "if ((js.Tick1Boo ? js.Tick1++ : js.Tick1--) >= 20) {js.Tick1Boo = false;} else if (js.Tick1 <= 0) {js.Tick1Boo = true;js.Tick1Angle = (js.Tick1Angle += 7) >= 360 ? 0 : js.Tick1Angle;var v = js.drawRound(js.Tick1Angle, 2)player.getWorld().spawnParticle(Particle.SPELL_WITCH, player.getLocation().clone().add(-v[0], 0.3 + js.Tick1 / 30.0, -v[1]), 1, 0, 0, 0, 0)player.getWorld().spawnParticle(Particle.SPELL_MOB, player.getLocation().clone().add(v[0], 0.5 + js.Tick1 / 20.0, v[1]), 1, 0, 0, 0, js.Tick1 / 100)"));
    config.set("List.Tick1.CombatPower", 20);
    return config;
}

/**
 * 启动方法
 */
function onEnable() {
    config.getConfigurationSection("List").getKeys(false).forEach(function (key) {
        JSRunnableList.push(JSRunnableManager.create(key));
    });
    JSAttribute.setLength(JSRunnableList.length);
}

/**
 * 重载方法
 */
function onReLoad() {
    JSRunnableList.forEach(function (jsRunnable) {
        jsRunnable.load();
    })
}


/**
 * 执行伤害/更新事件方法
 */
// eventMethod_debug = true
function eventMethod(values, eventData) {
    if (eventData.getClass().getSimpleName() === "UpdateData" && Player.isInstance(eventData.getEntity())) {
        var player = eventData.getEntity();
        for (var i = 0; i < JSRunnableList.length; i++) {
            var jsRunnable = JSRunnableList[i];
            if (jsRunnable.players.contains(player.getName())) {
                if (values[i] === 0) {
                    jsRunnable.disable(player);
                }
            } else {
                if (values[i] > 0) {
                    jsRunnable.enabled(player);
                }
            }
        }
    }

}

/**
 * 获取变量
 */
function getPlaceholder(values, player, string) {
    for (var i = 0; i < JSRunnableList.length; i++) {
        if (JSRunnableList[i].name === string) {
            return values[i];
        }
    }
}

function getPlaceholders() {
    var list = [];
    JSRunnableList.forEach(function (value, index) {
        list[index] = value.name;
    });
    return list;
}

function loadAttribute(values, lore) {
    for (var i = 0; i < JSRunnableList.length; i++) {
        if (lore.contains(JSRunnableList[i].discernName)) {
            values[i] += 1;
        }
    }
}

function calculationCombatPower(values) {
    var value = 0.0;
    for (var i = 0; i < JSRunnableList.length; i++) {
        if (values[i] > 0) {
            value += JSRunnableList[i].combatPower;
        }
    }
    return value;
}

